### PR TITLE
Add benchmarks using ASV

### PR DESCRIPTION
**Pull request by pv**
Convert the vbench suite from https://github.com/yarikoptic/numpy-vbench to [asv](http://asv.readthedocs.org/) and add it to Numpy repo with some helper commands added to runtests.

### CODE REVIEW

**Comment by charris** [comment]
Cool. Looks like six is a dependency, should we include it in numpy? How does the `--bench-compare` argument work?

**Comment by charris** [comment]
Also needs a mention in the 1.11 release notes.

**Comment by pv** [comment]
six is only required to get the `exec foo in ns` syntax in the benchmark
suite. It can't be included into numpy, because the benchmark suite
needs to be able to run against old numpy versions.
.
python runtests.py --bench-compare v1.6.2 bench_core

**Comment by charris** [comment]
There are some long lines that can be broken. Would it be worth letting the `--bench-compare` values be a comma separated list, or is it better to just use asv directly for that sort of thing.

**Comment by pv** [comment]
Changed to comma separated list.
Using asv directly would also work OK, but this saves you from typing
`git rev-parse HEAD`

**Comment by juliantaylor** [comment]
can you rerun benchmarks multiple times and take the minimum to reduce the spread?

**Comment by pv** [comment]
No builtin support for that, if you don't count what the timeit code
does. It's just a question of postprocessing json files, if you want to
hack it together on top of that.

File: benchmarks/benchmarks/bench_app.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    def setup(self, update):
+        N = 150
+        Niter = 1000
+        dx = 0.1
+        dy = 0.1
+        dx2 = (dx * dx)
+        dy2 = (dy * dy)
+
+        def num_update(u, dx2, dy2):
+            u[1:(-1), 1:(-1)] = ((((u[2:, 1:(-1)] + u[:(-2), 1:(-1)]) * dy2) + ((u[1:(-1), 2:] + u[1:(-1), :(-2)]) * dx2)) / (2 * (dx2 + dy2)))
```
Comment (charris) [inline benchmarks/benchmarks/bench_app.py]: Line too long.

File: runtests.py [hunk 2]
```diff
@@ -116,6 +126,9 @@ def main(argv):
     if args.gcov:
         gcov_reset_counters()
 
+    if args.debug and args.bench:
+        print("*** Benchmarks should not be run against debug version; remove -g flag ***")
```
Comment (charris) [inline runtests.py]: Too long?

File: runtests.py [hunk 3]
```diff
@@ -168,6 +181,57 @@ def main(argv):
...
+            else:
+                p.error("Too many commits to compare benchmarks for")
+
+            # Check for uncommitted files
+            if commit_b == 'HEAD':
+                r1 = subprocess.call(['git', 'diff-index', '--quiet', '--cached', 'HEAD'])
+                r2 = subprocess.call(['git', 'diff-files', '--quiet'])
+                if r1 != 0 or r2 != 0:
+                    print("*"*80)
+                    print("WARNING: you have uncommitted changes --- these will NOT be benchmarked!")
```
Comment (charris) [inline runtests.py]: Too long.

**Comment by pv** [comment]
Whitespaces fixed.

**Comment by charris** [comment]
Thanks Pauli.

**Pull request merged**